extensions/wikia/FilePage/WikiaFilePage.php

```diff
@@ -40,29 +40,96 @@ protected function isVideo() {
 	/**
 	 * Render the image or video
 	 */
-	protected function openShowImage(){
+	protected function openShowImage() {
 		if ( $this->isVideo() ) {
 			$this->openShowVideo();
 		} else {
 			parent::openShowImage();
 		}
 	}
 
+	/**
+	 * Render the image or video
+	 */
+	public function view() {
+		global $wgMemc, $wgDBprefix, $wgOut, $wgUser, $wgGroupPermissionsLocal, $wgNamespaceProtection;
+		if ( !$wgUser->isAnon() ) {
+			parent::view();
+
+			return;
+		}
+		//fallback to main page
+		$url = wfExpandUrl( "/wiki/Main_Page" );
+		//wiki needs read privileges
+		if ( preg_match( "/\*\|read\|0/", $wgGroupPermissionsLocal ) ||
+			 preg_match( "/\user\|read\|0/", $wgGroupPermissionsLocal ) ) {
+			$wgOut->redirect( $url );
+
+			return;
+		}
+		$redirKey = "redir:" . $wgDBprefix . ":" . $this->getTitle();
```

Comment:
I'd use `wfMemcKey` to build this, since then that makes sure the key is varied on environment and per wiki for you. I'd also use `$this->getTitle()->getPrefixedText()` rather than relying on `toString`.